core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/loadbalancer/EndpointGatewayNetworkValidator.java

```diff
@@ -0,0 +1,73 @@
+import com.sequenceiq.cloudbreak.validation.Validator;
+import com.sequenceiq.common.api.type.PublicEndpointAccessGateway;
+import com.sequenceiq.environment.api.v1.environment.model.response.EnvironmentNetworkResponse;
+
+@Component
+public class EndpointGatewayNetworkValidator implements Validator<Pair<String, EnvironmentNetworkResponse>> {
+
+    static final String NO_BASE_SUBNET = "No cluster subnet id specified. Endpoint gateway subnet cannot be determined.";
+
+    static final String NO_BASE_SUBNET_META = "Could not determine availability zone of cluster subnet %s";
+
+    static final String NO_USABLE_SUBNET_IN_ENDPOINT_GATEWAY = "Unable to find public subnet in availability zone %s in " +
+        "provided endoint gateway subnet list.";
+
+    static final String NO_USABLE_SUBNET_IN_CLUSTER = "Unable to find public subnet in availability zone %s in cluster subnets.";
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(EndpointGatewayNetworkValidator.class);
+
+    @Inject
+    private SubnetSelector subnetSelector;
+
+    @Override
+    public ValidationResult validate(Pair<String, EnvironmentNetworkResponse> subject) {
+        String baseSubnetId = subject.getLeft();
+        EnvironmentNetworkResponse network = subject.getRight();
+
+        ValidationResultBuilder resultBuilder = ValidationResult.builder();
+        if (network == null) {
+            LOGGER.debug("No network provided; public endpoint access gateway is disabled.");
+        } else {
+            if (PublicEndpointAccessGateway.ENABLED.equals(network.getPublicEndpointAccessGateway())) {
+                if (Strings.isNullOrEmpty(baseSubnetId)) {
+                    resultBuilder.error(NO_BASE_SUBNET);
+                } else {
+                    Optional<CloudSubnet> baseSubnet = subnetSelector.findSubnetById(network.getSubnetMetas(), baseSubnetId);
+                    if (baseSubnet.isEmpty()) {
+                        resultBuilder.error(String.format(NO_BASE_SUBNET_META, baseSubnetId));
+                    } else {
+                        if (!MapUtils.isEmpty(network.getGatewayEndpointSubnetMetas())) {
+                            LOGGER.debug("Attempting to validate endpoint gateway subnet using provided endpoint subnets.");
+                            Optional<CloudSubnet> endpointGatewaySubnet = subnetSelector.chooseSubnetForEndpointGateway(network, baseSubnetId);
+                            resultBuilder.ifError(endpointGatewaySubnet::isEmpty,
+                                String.format(NO_USABLE_SUBNET_IN_ENDPOINT_GATEWAY, baseSubnet.get().getAvailabilityZone()));
+                        } else {
+                            LOGGER.debug("Attempting to validate endpoint gateway subnet using cluster subnets.");
+                            Optional<CloudSubnet> endpointGatewaySubnet = subnetSelector.chooseSubnetForEndpointGateway(network, baseSubnetId);
```

Comment:
This statement is exactly the same one as the one in the `if` condition above. Seems fishy.

Reply:
The only reason it's broken out like this is so that when I set the error message, I can be very specific about which fields weren't set. I've refactored it a bit so that we only need to have the `subnetSelector.chooseSubnetForEndpointGateway` call once.